src/graph/graph.c

```diff
@@ -290,287 +124,291 @@ Graph *Graph_Get(RedisModuleCtx *ctx, RedisModuleString *graph_name) {
+            return i;
         }
     }
 
-    /* If access to newly created edges is requested
-     * reset iterator pass it back to caller. */
-    if(edges) {
-        *edges = iter;
-        DataBlockIterator_Reset(*edges);
+    // We should be able to find edge relation.
+    assert(false);
+}
+
+void Graph_GetEdgesConnectingNodes(const Graph *g, NodeID src, NodeID dest, int r, Edge **edges) {
+    assert(g && r < Graph_RelationTypeCount(g) && edges);
+
+    Edge e;
+    Entity *en;
+    Node srcNode;
+    Node destNode;
+    assert(Graph_GetNode(g, src, &srcNode));
+    assert(Graph_GetNode(g, dest, &destNode));
+    e.srcNodeID = src;
+    e.destNodeID = dest;
+    e.src = NULL;
+    e.dest = NULL;
+
+    if(r != GRAPH_NO_RELATION) {
+        en = _Graph_GetEdgeConnectingNodes(g, src, dest, r);
+        if(!en) return;
+        e.entity = en;
+        e.relationId = r;
+        *edges = array_append(*edges, e);
     } else {
-        DataBlockIterator_Free(iter);
+        // Relation type missing, scan through each edge type.
+        int relationCount = Graph_RelationTypeCount(g);
+        for(int i = 0; i < relationCount; i++) {
+            en = _Graph_GetEdgeConnectingNodes(g, src, dest, i);
+            if(!en) continue;
+            e.entity = en;
+            e.relationId = i;
+            *edges = array_append(*edges, e);
+        }
     }
 }
 
-Node* Graph_GetNode(const Graph *g, NodeID id) {
-    assert(g && id < Graph_NodeCount(g));
-    Node *n = (Node*)DataBlock_GetItem(g->nodes, id);
-    n->id = id;
-    return n;
-}
-
-Edge *Graph_GetEdge(const Graph *g, EdgeID id) {
-    assert(g && id < Graph_EdgeCount(g));
-    Edge *e = (Edge*)DataBlock_GetItem(g->edges, id);
-    e->id = id;
-    return e;
-}
+void Graph_CreateNode(Graph* g, int label, Node *n) {
+    assert(g);
 
-void Graph_GetEdgesConnectingNodes(const Graph *g, NodeID src, NodeID dest, int relation, Vector *edges) {
-    assert(g && src < Graph_NodeCount(g) && dest < Graph_NodeCount(g) && edges);
-    Edge *e;
-    EdgeDesc lookupKey = {src, dest, relation};
-    
-    // Search for edges.
-    if(relation != GRAPH_NO_RELATION) {
-        // Relation type specified.
-        HASH_FIND(hh, g->_edgesHashTbl, &lookupKey, sizeof(EdgeDesc), e);
-        if(e) Vector_Push(edges, e);
-    } else {
-        // Relation type missing, scan through each edge type.
-        for(int r = 0; r < g->relation_count; r++) {
-            // Update lookup relation id.
-            lookupKey.relationId = r;
-            // See if there's an edge of type 'r' connecting source to destination.
-            HASH_FIND(hh, g->_edgesHashTbl, &lookupKey, sizeof(EdgeDesc), e);
-            if(e) Vector_Push(edges, e);
+    NodeID id;
+    Entity *en = DataBlock_AllocateItem(g->nodes, &id);
+    en->id = id;
+    n->entity = en;
+
+    if(label != GRAPH_NO_LABEL) {
+        // Try to set matrix at position [id, id]
+        // incase of a failure, scale matrix.
```

Comment:
This is an interesting approach! I think I like it (though I'll take that back if it proves problematic).

Reply:
Haven't tested performance implications, but this would only work when our matrices have some buffer to them, or when we're reusing a deleted ID.